hircluster.c

```diff
@@ -3781,6 +3701,132 @@ static void redisClusterAsyncCallback(redisAsyncContext *ac, void *r,
     cluster_async_data_free(cad);
 }
 
+/* Reply callback function for CLUSTER SLOTS */
+void clusterSlotsReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    if (r == NULL) {
+        /* Ignore failures, next random node is hopefully working */
+        return;
+    }
+    redisReply *reply = (redisReply *)r;
+    dict *nodes = parse_cluster_slots(cc, reply, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+}
+
+/* Reply callback function for CLUSTER NODES */
+void clusterNodesReplyCallback(redisAsyncContext *ac, void *r, void *privdata) {
+    UNUSED(ac);
+    redisClusterAsyncContext *acc = (redisClusterAsyncContext *)privdata;
+    redisClusterContext *cc = acc->cc;
+    acc->update_route_time = hi_usec_now() + SLOTMAP_UPDATE_THROTTLE_USEC;
+
+    if (r == NULL) {
+        /* Ignore failures, next random node is hopefully working */
+        return;
+    }
+    redisReply *reply = (redisReply *)r;
+    dict *nodes = parse_cluster_nodes(cc, reply->str, reply->len, cc->flags);
+    if (updateNodesAndSlotmap(cc, nodes) != REDIS_OK) {
+        /* Ignore failures for now */
+    }
+}
+
+/* Return a random active node with async connection. */
+static redisClusterNode *getRandomActiveNode(dict *nodes) {
+    redisClusterNode *node;
+    dictIterator di;
+    dictInitIterator(&di, nodes);
```

Comment:
Can getRandomNode and RandomActiveNode be combined into one that prefers an active node? Traversing the nodes 4 times seems like a waste if there are 1000 nodes...

Perhaps it doesn't need to be completely fair either. We could for example pick one out of the first 10 nodes. Only if none of them is connected we can continue beyond that and pick the first connected node.

Reply:
The selection is now reworked.
1 of the 4 first found connected nodes are chosen, and if no connected nodes are found the last node that has not been used within 1 second is used (i.e a connection has not been attempted within 1 sec)